src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1219,18 +1220,77 @@ bool isMetadataOnlyConversion(const IDataType * from, const IDataType * to)
     }
 }
 
+/// Conversion that is allowed for partition key.
+/// Partition key should be serialized in the same way after conversion.
+/// NOTE: The list is not complete.
+/// TODO: Provide generic code.
+bool isSafeForPartitionKeyConversion(const IDataType * from, const IDataType * to)
+{
+    if (from->getName() == to->getName())
+        return true;
+
+    /// Enums are serialized in partition key as numbers - so conversion from Enum to number is Ok.
+    /// But not from number to Enum because Enum does not necessarily represents all numbers.
+    /// Numbers are serialized in text - so extending the type of number is Ok.
+    /// Conversion to/from Floats and DateTimes can also be allowed, but we don't enable it deliberately.
+
+    static const std::unordered_multimap<std::type_index, const std::type_info &> ALLOWED_CONVERSIONS =
+        {
+            { typeid(DataTypeEnum8),    typeid(DataTypeEnum8)  },
+            { typeid(DataTypeEnum8),    typeid(DataTypeEnum16) },
+            { typeid(DataTypeEnum8),    typeid(DataTypeInt8)   },
+            { typeid(DataTypeEnum8),    typeid(DataTypeInt16)  },
+            { typeid(DataTypeEnum8),    typeid(DataTypeInt32)  },
+            { typeid(DataTypeEnum8),    typeid(DataTypeInt64)  },
+
+            { typeid(DataTypeEnum16),   typeid(DataTypeEnum16) },
+            { typeid(DataTypeEnum16),   typeid(DataTypeInt16)  },
+            { typeid(DataTypeEnum16),   typeid(DataTypeInt32)  },
+            { typeid(DataTypeEnum16),   typeid(DataTypeInt64)  },
```

Comment:
Extendenting the size of column (like from UInt8 to uint64), will also require refreshing `minmax_` file. Is it recounted now on ALTERs automatically?

Reply:
Yes, seems like we will not be able to deserialize partition key minmax from disk https://github.com/clickhouse/ClickHouse/blob/master/src/Storages/MergeTree/IMergeTreeDataPart.cpp#L44-L55. Also we have a separate method for internal structures update https://github.com/clickhouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeData.cpp#L418.